python/pyspark/sql/connect/column.py

```diff
@@ -314,3 +323,62 @@ def to_plan(self, session: "RemoteSparkSession") -> proto.Expression:
 
     def __str__(self) -> str:
         return f"({self._op} ({', '.join([str(x) for x in self._args])}))"
+
+
+class Column(object):
```

Comment:
Renaming `Expression` to `Column` would break `type(expr) is Column`. This means that existing code has to be rewritten to `isinstance(expr, Column)`. Nesting the expression seems like the better option here, given that the goal of these changes is API compatibility.

Reply:
I think that is an interesting observation, not sure though how much this is a relevant case in practice because `type` explicitly does not check inheritance and is usually much to fragile.

Second, I'm not sure that the choice was actually because of this particularity in Python.

Reply:
@HyukjinKwon any though on this?

Reply:
In Catalyst, we have to wrap `Expression` with `Column`, because `Expression` is an internal class with many internal APIs and we can't expose it to end users directly.

In Spark Connect, I think we can skip the wrapping if `Expression` doesn't have any internal APIs. But end users will know we have different concrete `Column` subclasses, which is kind of an internal detail. If this doesn't matter, I think `Column` and `Expression` can be merged. cc @grundprinzip

Reply:
This is my question given I am not familiar with PySpark:

If we have a base class `Column` and then have many sub-class inherits it to form the Expression family, does it count as leaking internal implementation to users for PySpark or Python users when they start to use Spark Connect Python client?

Reply:
If we're merging `Expression` to `Column`, we should remove all these `SortOrder`, etc (which does not exist in PySpark). For the current hierarchy, it looks fine to wrap `Expression` with `Column` since `Expression`s here look more internal anyway

Reply:
@grundprinzip can you give your opinion clearly? It seems most reviewers here agree to have `Column` wrapping `Expression`, as it can hide the implementation details the most. It's also consistent with `DataFrame` wrapping `LogicalPlan` in Spark Connect today. Are you OK with moving forward with this design?

Reply:
I stated already that I'm fine with wrapping expressions. My previous comment was on the sort order.

Reply:
OK cool, now this PR is unblocked. @amaliujia can you resolve the code conflicts, please?

Reply:
Thank all for this discussion! I will rebase and resolve conflicts.